### DISCUSSION TITLE
Modrinth publishing error: "GitHub Token from incorrect Client ID"

### DISCUSSION THREAD

**Discussion by srnyx** [Kira-NT/mc-publish#134] [Q&A] [answered]
I get this error only when publishing to Modrinth:
`Error: HttpError: 401 (Unauthorized, {"error":"unauthorized","description":"Authentication Error: GitHub Token from incorrect Client ID"})`

**Job:** https://github.com/srnyx/critical-colors/actions/runs/11759568938/job/32759169521

CurseForge and GitHub publishing works fine for some reason, no idea why it's just Modrinth.

**Comment by Kira-NT** [comment, marked as answer]
> CurseForge and GitHub publishing works fine for some reason, no idea why it's just Modrinth.

Because CurseForge and GitHub have nothing to do with the Modrinth token, thus they do not fail :D

You may be trying to use an outdated token *(Modrinth's folks changed how tokens are issued quite some time ago)*, or your token may have expired *(in which case the error message could use some work)*, or there could have been recent backend changes to Modrinth's authorization system as a whole.

In any case, simply issue a new Modrinth token for yourself, and you should be good to go!

**Reply by srnyx** [reply]
> Because CurseForge and GitHub have nothing to do with the Modrinth token, thus they do not fail :D

But the error message says `GitHub Token from incorrect Client ID`, nothing about the Modrinth token. I'm just not sure what GitHub token it is referring to and why it matters when it's posting to Modrinth.

I'll do as you suggested anyways though, as it's definitely possible I forgot to update the token for some of my projects and the error message is just not being descriptive or something. Will let you know how it goes.

**Reply by Kira-NT** [reply]
> I'm just not sure what GitHub token it is referring to and why it matters when it's posting to Modrinth.

Modrinth uses GitHub as one of its authentication providers, which is why you're seeing this error message. This makes perfect sense if you're familiar with what Modrinth packs under the hood, but for end users, it's a completely different story. So, you're not the first *(and certainly won't be the last)* to be confused by these error messages.

[Here's](https://github.com/modrinth/code/blob/3fa07f64d381ca62ec15ac2c126e42625c0f76d2/apps/labrinth/src/auth/mod.rs#L44-L45) where the error you're encountering is defined:

```rust
#[error("GitHub Token from incorrect Client ID")]
InvalidClientId,
```

And [here's](https://github.com/modrinth/code/blob/3fa07f64d381ca62ec15ac2c126e42625c0f76d2/apps/labrinth/src/routes/internal/flows.rs#L573-L577) where it gets thrown:

```rust
if client_id
    != Some(&*dotenvy::var("GITHUB_CLIENT_ID").unwrap())
{
    return Err(AuthenticationError::InvalidClientId);
}
```

Judging by the code, it looks like you're using a `gho_*` token. These tokens have been deprecated for quite some time now *(honestly, it feels like it's been a few years at this point, though I might be mistaken)*, so it's **definitely** time to refresh yours :D
Modern Modrinth tokens follow a different format: `mrp_*`.

**Reply by srnyx** [reply]
> Judging by the code, it looks like you're using a `gho_*` token. These tokens have been deprecated for quite some time now _(honestly, it feels like it's been a few years at this point, though I might be mistaken)_, so it's **definitely** time to refresh yours :D Modern Modrinth tokens follow a different format: `mrp_*`.

Think you're right, my secret on GitHub was last updated July 31, 2023 but I made a new Modrinth token September 28, 2023. So I just forgot to update it for all my repositories. Thanks.
*Only been a year by the way since Modrinth's original tokens were deprecated I think*

**Reply by Kira-NT** [reply]
You are welcome, I'm always glad to help!

> Only been a year by the way since Modrinth's original tokens were deprecated I think

That's still a pretty reasonable timeframe to expect the users to be done with refreshing their tokens! :D
